internal/pkg/deploy/upload/asset/asset.go

```diff
@@ -71,98 +89,73 @@ func walkFn(source, dest string, recursive bool, reader afero.Fs, upload UploadF
+			dest = info.Name()
+		}
 
-func buildReincludeMatchers(reincludes []string) []filepathMatcher {
-	var matchers []filepathMatcher
-	for _, reinclude := range reincludes {
-		matchers = append(matchers, reincludeMatcher(reinclude))
+		*assets = append(*assets, asset{
+			localPath:          path,
+			content:            buf,
+			ArtifactBucketPath: filepath.Join(u.PathPrefix, hex.EncodeToString(hash.Sum(nil))),
+			ServiceBucketPath:  dest,
+		})
+		return nil
 	}
-	return matchers
-}
-
-func (m reincludeMatcher) match(path string) (bool, error) {
-	return match(string(m), path)
 }
 
-type excludeMatcher string
+func (u *ArtifactBucketUploader) uploadAssets(assets []asset) error {
+	g, _ := errgroup.WithContext(context.Background())
 
-func buildExcludeMatchers(excludes []string) []filepathMatcher {
-	var matchers []filepathMatcher
-	for _, exclude := range excludes {
-		matchers = append(matchers, excludeMatcher(exclude))
+	for i := range assets {
+		asset := assets[i]
+		g.Go(func() error {
+			if err := u.Upload(asset.ArtifactBucketPath, asset.content); err != nil {
+				return fmt.Errorf("upload %q: %w", asset.localPath, err)
+			}
+			return nil
+		})
 	}
-	return matchers
-}
 
-func (m excludeMatcher) match(path string) (bool, error) {
-	return match(string(m), path)
-}
-
-// compositeMatcher is a composite matcher consisting of reinclude matchers and exclude matchers.
-// Note that exclude matchers will be applied before reinclude matchers.
-type compositeMatcher struct {
-	reincludeMatchers []filepathMatcher
-	excludeMatchers   []filepathMatcher
-}
-
-func buildCompositeMatchers(reincludeMatchers, excludeMatchers []filepathMatcher) compositeMatcher {
-	return compositeMatcher{
-		reincludeMatchers: reincludeMatchers,
-		excludeMatchers:   excludeMatchers,
-	}
+	return g.Wait()
 }
 
-func (m compositeMatcher) match(path string) (bool, error) {
-	shouldInclude := true
-	for _, matcher := range m.excludeMatchers {
-		isMatch, err := matcher.match(path)
-		if err != nil {
-			return false, err
-		}
-		if isMatch {
-			shouldInclude = false
-		}
-	}
-	for _, matcher := range m.reincludeMatchers {
-		isMatch, err := matcher.match(path)
-		if err != nil {
-			return false, err
+func (u *ArtifactBucketUploader) uploadAssetMappingFile(assets []asset) error {
+	// stable output
+	sort.Slice(assets, func(i, j int) bool {
+		if assets[i].ArtifactBucketPath != assets[j].ArtifactBucketPath {
+			return assets[i].ArtifactBucketPath < assets[j].ArtifactBucketPath
 		}
-		if isMatch {
-			shouldInclude = true
-		}
-	}
-	return shouldInclude, nil
-}
+		return assets[i].ServiceBucketPath < assets[j].ServiceBucketPath
+	})
 
-func match(pattern, path string) (bool, error) {
-	isMatch, err := filepath.Match(pattern, path)
+	data, err := json.Marshal(assets)
 	if err != nil {
-		return false, fmt.Errorf("match file path %s against pattern %s: %w", path, pattern, err)
+		return fmt.Errorf("encode uploaded assets: %w", err)
```

Comment:
```suggestion
		return fmt.Errorf("encode staged assets: %w", err)
```
would this be more accurate? Same for the similar situation otherwise users would think their file has already been "uploaded".

Reply:
actually at this point their files have been uploaded to the cache bucket